server/data/data-jmap/src/main/java/org/apache/james/jmap/api/upload/UploadServiceDefaultImpl.java

```diff
@@ -90,36 +88,26 @@ private Mono<Void> cleanupUpload(Username username, long currentUploadUsage) {
         long minimumSpaceToClean = currentUploadUsage - jmapUploadQuotaConfiguration.getUploadQuotaLimitInBytes() / 2;
         AtomicLong cleanedSpace = new AtomicLong(0L);
 
-        if (shouldPerformReadRepair()) {
-            return Flux.from(uploadRepository.listUploads(username))
-                .sort(Comparator.comparing(UploadMetaData::uploadDate))
-                .flatMap(upload -> {
-                    if (cleanedSpace.get() < minimumSpaceToClean) {
-                        return Mono.from(uploadRepository.delete(upload.uploadId(), username))
-                            .then(Mono.from(uploadUsageRepository.decreaseSpace(username, QuotaSizeUsage.size(upload.sizeAsLong()))))
-                            .then(Mono.fromCallable(() -> cleanedSpace.addAndGet(upload.sizeAsLong())))
-                            .then(Mono.empty());
-                    } else {
-                        return Mono.fromCallable(() -> upload);
-                    }
-                }, LOW_CONCURRENCY)
-                .map(UploadMetaData::sizeAsLong)
-                .reduce(Long::sum)
-                .filter(totalSpaceUsed -> totalSpaceUsed != currentUploadUsage)
-                .flatMap(totalSpaceUsed -> Mono.from(uploadUsageRepository.resetSpace(username,
-                    QuotaSizeUsage.size(totalSpaceUsed))))
-                .then();
-        } else {
-            return Flux.from(uploadRepository.listUploads(username))
-                .sort(Comparator.comparing(UploadMetaData::uploadDate))
-                .takeWhile(upload -> cleanedSpace.addAndGet(upload.sizeAsLong()) <= minimumSpaceToClean)
-                .flatMap(upload -> Mono.from(uploadRepository.delete(upload.uploadId(), username))
-                    .then(Mono.from(uploadUsageRepository.decreaseSpace(username, QuotaSizeUsage.size(upload.sizeAsLong())))))
-                .then();
-        }
+        return Flux.from(uploadRepository.listUploads(username))
+            .sort(Comparator.comparing(UploadMetaData::uploadDate))
+            .flatMap(deleteUpload(username, minimumSpaceToClean, cleanedSpace), LOW_CONCURRENCY)
+            .map(UploadMetaData::sizeAsLong)
+            .reduce(Long::sum)
+            .filter(totalSpaceUsed -> totalSpaceUsed != currentUploadUsage)
+            .flatMap(totalSpaceUsed -> Mono.from(uploadUsageRepository.resetSpace(username, QuotaSizeUsage.size(totalSpaceUsed))))
```

Comment:
IMO it will be better if call `resetSpace` only one time after the all deleteUpload

Reply:
The current code should already "call resetSpace only one time after the all deleteUpload". cf: a lot of `.then` (after `.delete`) and a final `.reduce`.